src/ir/proto_to_ir.h

```diff
@@ -0,0 +1,122 @@
+  ProtoToIR& operator=(const ProtoToIR&) = delete;
+
+  virtual ~ProtoToIR() {}
+
+  static Result Convert(IRContext& context,
+                        const proto::IrTranslationUnit& root) {
+    // Store the operators in the context.
+    for (const auto& op : root.operators()) {
+      auto name = op.name();
+      // Register new operators.
+      if (!context.IsRegisteredOperator(name)) {
+        Operator op = Operator(name);
+        context.RegisterOperator(std::make_unique<Operator>(std::move(op)));
+      }
+    }
+
+    auto ssa_names = std::make_unique<SsaNames>();
+    ProtoToIR state(context, *ssa_names);
+    // Construct objects to point to during construction.
+    auto module = state.PreVisit(root.top_level_module());
+    // Perform construction.
+    state.ToIR(root.top_level_module());
+    // Move objects into their 'owning' objects.
+    state.PostVisit(*module, root.top_level_module());
+
+    // TODO: Check that the construction correctly handled ownership of all
+    // referenced objects. Currently empty builders are left behind... filter?
+    for(const auto& block_it : state.blocks_) {
+      CHECK(block_it.second.IsBuilt()) << "Block " << block_it.first << " was not built and placed in it's module.";
+    }
+    for(const auto& operation_it : state.operations_) {
+      CHECK(operation_it.second.get() == nullptr) << "Operation " << operation_it.first << " was not moved into it's module.";
+    }
+    return Result{
+      .module = std::move(module),
+      .ssa_names = std::move(ssa_names)
+    };
+  }
+
+  std::unique_ptr<Module> PreVisit(const proto::Module& module_proto);
+  void PreVisit(const proto::Block& block_proto);
+  void PreVisit(const proto::Operation& operation_proto);
+  void PreVisit(const proto::Value& value);
+  void PreVisit(const proto::AnyValue& any_value);
+  void PreVisit(const proto::BlockArgumentValue& block_argument);
+  void PreVisit(const proto::OperationResultValue&);
+  void PreVisit(const proto::AttributePayload&);
+
+  void ToIR(const proto::Module& module_proto);
+  void ToIR(ID block_id, const proto::Block& block_proto);
+  void ToIR(Operation& op, const proto::Operation& operation_proto);
+  Value ToIR(const proto::Value& value);
+  value::Any ToIR(const proto::AnyValue&);
+  value::BlockArgument ToIR(const proto::BlockArgumentValue& block_argument);
+  value::OperationResult ToIR(const proto::OperationResultValue&);
+  Attribute ToIR(const proto::AttributePayload&);
+
+  void PostVisit(Module& module, const proto::Module& module_proto);
+  void PostVisit(BlockBuilder& block, const proto::Block& block_proto);
+  void PostVisit(Operation& op, const proto::Operation& operation_proto);
+  void PostVisit(const proto::Value& value);
+  void PostVisit(const proto::AnyValue& any_value);
+  void PostVisit(const proto::BlockArgumentValue& block_argument);
+  void PostVisit(const proto::OperationResultValue&);
+  void PostVisit(const proto::AttributePayload&);
```

Comment:
Shoudln't all of these methods be private?

Reply:
Actually probably! Moved as much as I could to the `private:` section.